integration_tests/src/main/python/array_test.py

```diff
@@ -625,3 +625,37 @@ def test_arrays_overlap_before_spark313(data_gen):
             'arrays_overlap(array(3, 4), array(1, 2))',
             'arrays_overlap(array(), array(1, 2))')
     )
+
+@pytest.mark.parametrize('data_gen', [byte_gen, short_gen, int_gen, long_gen,
+                                      float_gen, double_gen,
+                                      string_gen, boolean_gen, date_gen, timestamp_gen] + decimal_gens, ids=idfn)
+@pytest.mark.skipif(is_before_spark_313(), reason="NaN equality is only handled in Spark 3.1.3+")
+def test_array_remove(data_gen):
+    gen = StructGen(
+        [('a', ArrayGen(data_gen, nullable=True)),
+         ('b', data_gen)],
+         nullable=False)
+
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: gen_df(spark, gen).selectExpr(
+            'array_remove(a, b)',
+            'array_remove(array(1, 2, 2, 3, null), null)',
+            'array_remove(array(1, 2, 2, 3, null), 2)')
+    )
+
+@pytest.mark.parametrize('data_gen', [byte_gen, short_gen, int_gen, long_gen,
+                                      FloatGen(no_nans=True), DoubleGen(no_nans=True),
+                                      string_gen, boolean_gen, date_gen, timestamp_gen] + decimal_gens, ids=idfn)
+@pytest.mark.skipif(not is_before_spark_313(), reason="NaN equality is only handled in Spark 3.1.3+")
```

Comment:
Do we plan on being bug for bug compatible with Spark for 3.1.2 and below? or do we just plan on documenting it? Do you have the issue in Spark that made the change? Just curious if it was an intentional bug fix or if it was a side effect of a different change.

Reply:
This looks like it was based on the previous integration tests that were testing the set-based operations (array_intersect, array_union, array_except, etc.).  In those particular cases, we did have compatibility issues with Spark 3.1.2 and below (which was due to the hashCode implementation in Apache Spark in those cases that uses the binary representation of double/float). Can we confirm that it would specifically manifest in `array_remove` as well?

Reply:
I looked into this and the compatibility issues were due to `OpenHashSet` handling `NaN` values. `array_remove` does not use `OpenHashSet` implementation and should not have these problems. I have updated the tests. Thank you very much for pointing this out.